driver-core/src/test/unit/com/mongodb/internal/connection/ByteBufferBsonInputTest.java

```diff
@@ -0,0 +1,719 @@
+        // given
+        ByteBuf buffer = allocateAndWriteToBuffer(bufferProvider, nonNullTerminatedCString);
+        try (ByteBufferBsonInput bufferInput = new ByteBufferBsonInput(buffer)) {
+
+            // when & then
+            assertThrows(BsonSerializationException.class, bufferInput::readCString);
+        }
+    }
+
+
+    @ParameterizedTest
+    @MethodSource("bufferProviders")
+    void shouldThrowIfOneByteStringIsNotNullTerminated(final BufferProvider bufferProvider) {
+        // given
+        ByteBuf buffer = allocateAndWriteToBuffer(bufferProvider, new byte[]{2, 0, 0, 0, 1});
+        try (ByteBufferBsonInput bufferInput = new ByteBufferBsonInput(buffer)) {
+
+            // when & then
+            assertThrows(BsonSerializationException.class, bufferInput::readString);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("bufferProviders")
+    void shouldThrowIfOneByteCStringIsNotNullTerminated(final BufferProvider bufferProvider) {
+        // given
+        ByteBuf buffer = allocateAndWriteToBuffer(bufferProvider, new byte[]{1});
+        try (ByteBufferBsonInput bufferInput = new ByteBufferBsonInput(buffer)) {
+
+            // when & then
+            assertThrows(BsonSerializationException.class, bufferInput::readCString);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("bufferProviders")
+    void shouldThrowIfLengthOfBsonStringIsNotPositive(final BufferProvider bufferProvider) {
+        // given
+        ByteBuf buffer = allocateAndWriteToBuffer(bufferProvider, new byte[]{-1, -1, -1, -1, 41, 42, 43, 0});
+        try (ByteBufferBsonInput bufferInput = new ByteBufferBsonInput(buffer)) {
+
+            // when & then
+            assertThrows(BsonSerializationException.class, bufferInput::readString);
+        }
+    }
+
+    public static Stream<Arguments> shouldSkipCStringWhenMultipleNullTerminationPresent() {
+        List<Arguments> arguments = new ArrayList<>();
+        List<BufferProvider> collect = bufferProviders().collect(toList());
+        for (BufferProvider bufferProvider : collect) {
+            arguments.add(Arguments.of(new byte[]{0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0x4b, 0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0x4b, 0x4c, 0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0x61, 0x76, 0x61, 0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0x61, 0x76, 0x61, 0x62, 0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0x61, 0x76, 0x61, 0x65, 0x62, 0x67, 0, 8, 0, 0, 0}, bufferProvider));
+            arguments.add(Arguments.of(new byte[]{0x4a, 0, 8, 0, 0, 0}, bufferProvider));
+        }
```

Comment:
If you feel tired with such manual cartesian product generation (2 dimentions are still manageable), you can try the `JUnit Pioneer` library, which Spring Data Mongo project is using: https://junit-pioneer.org/docs/cartesian-product/.

Reply:
this message is purely for discussion purpose. Closed it for now.